torch/csrc/distributed/rpc/request_callback.cpp

```diff
@@ -20,9 +20,27 @@ Message createException(const Message& request, const std::exception& e) {
       request.id());
 }
 
+// When request message has autograd info, processMessage() will set up valid
+// current context id properly. This struct is used to clean up current context
+// id after processMessage() is done.
+struct ClearAutogradContextGuard {
+  ClearAutogradContextGuard() = default;
+  ~ClearAutogradContextGuard() {
+    clear();
+  }
+
+  void clear() {
+    auto& autogradContainer = DistAutogradContainer::getInstance();
+    autogradContainer.clearCurrentContext();
+  }
+};
+
 } // anonymous namespace
 
 Message RequestCallback::operator()(Message& request) const {
+  // For a rev thread, current context id should be invalid outside
+  // processMessage().
+  ClearAutogradContextGuard guard;
```

Comment:
(does not need to be done in this PR) It looks a little weird that we have a guard that only clears context but not set it. I wonder if it would be better if we let the guard to govern both set and clear context?

Reply:
@mrshenli this is possibly hard to set context here, as we can only set it after addRecvBackward() in FORWARD_AUTOGRAD_REQ